### PR TITLE
Fix a bug with  metric queries and label_format.

### PR DESCRIPTION
**Pull request by cyriltovena**
In my previous PR where I introduced hints, I forgot to implement label format label required.
This would cause label format label to be skipped during parsing.

I've also took the opportunity to improve the code.

Signed-off-by: Cyril Tovena <cyril.tovena@gmail.com>

### CODE REVIEW

File: pkg/logql/log/metrics_extraction.go [hunk 1]
```diff
@@ -226,3 +230,75 @@ func convertBytes(v string) (float64, error) {
...
+// ParserHint are hints given to LogQL parsers.
+// This is specially useful for parser that extract implicitly all possible label keys.
+// This is used only within metric queries since it's rare that you need all label keys.
+// For example in the following expression:
+//
+//		sum by (status_code) (rate({app="foo"} | json [5m]))
+//
+// All we need to extract is the status_code in the json parser.
+type ParserHint interface {
+	// Tells is a label with the given key should be extracted.
```
Comment (kavirajk) [inline pkg/logql/log/metrics_extraction.go:243]:
```suggestion
	// Tells if a label with the given key should be extracted.
```

File: pkg/logql/log/metrics_extraction.go [hunk 2]
```diff
@@ -226,3 +230,75 @@ func convertBytes(v string) (float64, error) {
...
+func (p *parserHint) NoLabels() bool {
+	return p.noLabels
+}
+
+// newParserHint creates a new parser hint using the list of labels that are seen and required in a query.
+func newParserHint(requiredLabelNames []string, without, noLabels bool, metricLabelName string) *parserHint {
+	if noLabels {
+		// if we need to extract metric from a label, then use that as required labels list.
+		if metricLabelName != "" {
+			return &parserHint{requiredLabels: []string{metricLabelName}}
```
Comment (kavirajk) [inline pkg/logql/log/metrics_extraction.go:294]: Q: wondering why we need `metricLabelName` as separate param? whoever using parserHint already knows wether its metricLabel or not correct? and they can pass it directly as `requiredLabelNames` in that case? No?

I'm thinking why not something like this?

```
if noLabels {
     return &parserHint{noLabels: true}
}

if without {
		return noParserHints
}

return &parserHint{requiredLabels: requiredLabelNames}
```

Basically I'm thinking is the combination of `noLabels` and only `metricLabelName` is needed? at least it sounds contradicting with the naming.

Comment (owen-d) [reply pkg/logql/log/metrics_extraction.go:294]: Consider
```
sum(
  sum_over_time(
  {job="loki-prod/query-frontend"}
    |= "metrics.go:83"
    | logfmt
    | unwrap duration(duration)
    [10m]
  )
)
```

In this case, we're sum'ing across all series, looking for an aggregated result with the labels `{}`. However, we _need_ to extract the `duration` field used in the extract, which is the `metricLabelName` here.

**Review by owen-d — APPROVED** [review]
LGTM

**Comment by cyriltovena** [comment]
> LGTM

Made some mistake will add more tests in the logql package to ensure this optimization is not breaking anything it's the second time and I think it's because I don't have a good test story.

**Pull request merged**